README.md

```diff
@@ -1,486 +1,679 @@
+export function defineElement(name, options) {
+  return klass => { customElements.define(name, klass, options); return klass; }
 }
 ```
 
-This example uses a `decorator` declaration with a parameter list, which lets a decorator definition take arguments that can be used to supply arguments to other decorators in its definition.
+The decorator takes arguments at its usage site, so it is implemented as a function that returns another function. You can think of it as a "decorator factory": after you apply the arguments, it gives you another decorator.
 
-### `@metadata`
+This decorator usage could be desugared as follows:
 
-The `@metdata(key, value)` decorator is similar to [`@Reflect.metadata`](https://github.com/rbuckton/reflect-metadata): It allows the easy retrieval of information which was stored by annotating the class. The following examples are written in terms of the Reflect.metadata proposal, but you could imagine storing the metadata in any other form.
+```js
+class MyClass extends HTMLElement { }
+MyClass = defineElement('my-class')(MyClass, {kind: "class"});
+```
 
-```mjs
-import { @metadata } from "./metadata.mjs";
+### Decorators adding metadata
 
-// partially apply the decorator locally for terseness
-decorator @localMeta { @metadata("key", "value") }
+Decorators can add metadata about class elements by adding a `metadata` property of the context object that is passed in to them. All of the metadata objects are `Object.assign`'ed together and placed in a property reachable from `[Symbol.metadata]` on the class. For example:
 
-@localMeta class C {
-  @localMeta method() { }
+```mjs
+@annotate({x: "y"}) @annotate({v: "w"}) class C {
+  @annotate({a: "b"}) method() { }
+  @annotate({c: "d"}) field;
 }
 
-Reflect.getMetadata(C, "key");                      // "value"
-Reflect.getMetadata(C.prototype, "key", "method");  // "value"
+C[Symbol.metadata].class.x                     // "y"
+C[Symbol.metadata].class.v                     // "w"
+C[Symbol.metadata].prototype.methods.method.a  // "a"
+C[Symbol.metadata].instance.fields.field.c     // "d"
 ```
 
-`@metadata` can also be defined in terms of `@register`. When `@register` is used with a public field, method or accessor, it is called with the second argument being the property key of that class element.
+**NOTE:** The exact format of the annotations object is not very well-thought-out and could use more refinement. The main thing I'd like to illustrate here is, it's just an object, with no particular support library to read or write it, and it's automatically created by the system.
 
-```mjs
-// metadata.mjs
-import "reflect-metadata";
+This decorator `@annotate` could be implemented as follows:
 
-export decorator @metadata(key, value) {
-  @register((target, prop) => Reflect.defineMetadata(key, value, target, prop))
+```js
+function annotate(metadata) {
+  return (_, context) => {
+    context.metadata = metadata;
+    return _;
```

Comment:
could the return statement be omitted for the same effect?

Reply:
That's a decision we could still make; I was leaning towards requiring the return statement, but I see this as a detail we can refine after landing this patch.

Reply:
Totally fine to iterate later; I would personally find that anything side-effecting having a return value is unclear/confusing.

Reply:
This proposal specifically permits decorators to both wrap a class element *and* modify metadata.